### PR TITLE
[Macros] Plugin search options group

### PR DESCRIPTION
**Pull request by rintaro**
`load-plugin-library`, `load-plugin-executable`, `-plugin-path` and `-external-plugin-path` should be searched in the order they are specified in the arguments.

Previously, for example `-plugin-path` used to precede `-external-plugin-path` regardless of the position in the arguments.

### CODE REVIEW

File: include/swift/AST/SearchPathOptions.h [hunk 1]
```diff
@@ -34,7 +35,6 @@ enum class ModuleSearchPathKind {
   Framework,
   DarwinImplicitFramework,
   RuntimeLibrary,
-  CompilerPlugin,
```
Comment (rintaro) [inline include/swift/AST/SearchPathOptions.h:37]: `ModuleSearchPathKind::CompilePlugin` was incomplete (only set for `-load-plugin-library`), and was not actually used.

File: lib/AST/ASTContext.cpp [hunk 2]
```diff
@@ -103,11 +103,10 @@ llvm::StringRef swift::getProtocolName(KnownProtocolKind kind) {
 }
 
 namespace {
-  enum class SearchPathKind : uint8_t {
-    Import = 1 << 0,
-    Framework = 1 << 1,
-    CompilerPlugin = 1 << 2
```
Comment (rintaro) [inline lib/AST/ASTContext.cpp:109]: Ditto: `SearchPathKind::CompilerPlugin` wasn't actually used.

File: include/swift/AST/SearchPathOptions.h [hunk 3]
```diff
@@ -355,27 +367,6 @@ class SearchPathOptions {
     Lookup.searchPathsDidChange();
   }
 
-  void setCompilerPluginLibraryPaths(
-      std::vector<std::string> NewCompilerPluginLibraryPaths) {
-    CompilerPluginLibraryPaths = NewCompilerPluginLibraryPaths;
-    Lookup.searchPathsDidChange();
```
Comment (rintaro) [inline include/swift/AST/SearchPathOptions.h:361]: Plugin search doesn't use `Lookup`, and adding plugin search path won't affect `import` module search path. So we don't need to call `Lookup.searchPathsDidChange()` when updating `PluginSearchOpts`.

File: lib/AST/PluginLoader.cpp [hunk 4]
```diff
@@ -46,69 +48,59 @@ PluginRegistry *PluginLoader::getRegistry() {
...
-    llvm::sys::path::append(fullPath, expectedBasename);
-    if (fs->exists(fullPath)) {
-      return std::string(fullPath);
+  // FIXME: Shared library prefix might be different between platforms.
+  SmallString<64> pluginLibBasename;
+  pluginLibBasename.append("lib");
+  pluginLibBasename.append(moduleName.str());
+  pluginLibBasename.append(LTDL_SHLIB_EXT);
+
+  // FIXME: Should we create a lookup table keyed by module name?
```
Comment (rintaro) [inline lib/AST/PluginLoader.cpp:62]: Currently we create a lookup table for `-load-plugin-executable` only. But we could lazily create a lookup table for all keyed by module names. But that's need to scan (iterate directory entries) all search paths, so I'm not sure it worth it.

File: test/Macros/serialize_plugin_search_paths.swift [hunk 5]
```diff
@@ -15,7 +15,6 @@
 // CHECK:     -plugin-path: {{.*}}plugins
 // CHECK:     -plugin-path: {{.*}}plugins
 // CHECK:     -plugin-path: {{.*}}plugins
-// CHECK:     -plugin-path: {{.*}}plugins
```
Comment (rintaro) [inline test/Macros/serialize_plugin_search_paths.swift:18]: Previously manually specified `-plugin-path` was duplicated in `libDriver`.

File: lib/Frontend/Frontend.cpp [hunk 6]
```diff
@@ -209,30 +209,35 @@ SerializationOptions CompilerInvocation::computeSerializationOptions(
...
-  }
-  // '-load-plugin-executable' options.
-  for (const PluginExecutablePathAndModuleNames &pair :
-       getSearchPathOptions().getCompilerPluginExecutablePaths()) {
-    std::string optStr = pair.ExecutablePath + "#";
-    llvm::interleave(
-        pair.ModuleNames, [&](auto &name) { optStr += name; },
-        [&]() { optStr += ","; });
-    serializationOpts.CompilerPluginExecutablePaths.push_back(optStr);
+  // FIXME: Preserve the order of these options.
```
Comment (DougGregor) [inline lib/Frontend/Frontend.cpp:212]: Are you intentionally trying to avoid changing the module serialization format here?

**Review by DougGregor — APPROVED** [review]

File: lib/Frontend/Frontend.cpp (hunk 6 above)
Comment (rintaro) [reply lib/Frontend/Frontend.cpp:212]: Yes, my intension is doing it in a followup.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/Frontend/CompilerInvocation.cpp [hunk 7]
```diff
@@ -1601,18 +1601,56 @@ static bool ParseSearchPathArgs(SearchPathOptions &Opts,
...
+    }
+    case OPT_plugin_path: {
+      Opts.PluginSearchOpts.emplace_back(
+          PluginSearchOption::PluginPath{resolveSearchPath(A->getValue())});
+      break;
+    }
+    case OPT_external_plugin_path: {
+      // '<plugin directory>#<plugin server executable path>'.
+      // FIXME: '#' can be used in the paths.
+      StringRef dylibPath;
```
Comment (bnbarham) [inline lib/Frontend/CompilerInvocation.cpp:1644]: Not necessarily a dylib

File: test/Macros/macro_plugin_searchorder.swift [hunk 8]
```diff
@@ -50,39 +50,42 @@
...
+// RUN:   -o %t/main1
 // RUN: %target-codesign %t/main1
 // RUN: %target-run %t/main1 | %FileCheck --check-prefix=CHECK_LOAD_PLUGIN_LIBRARY %s
 
 //#-- Expect -load-plugin-executable
 // RUN: %target-build-swift %t/src/test.swift \
-// RUN:   -o %t/main2 \
 // RUN:   -module-name test \
+// RUN:   -load-plugin-executable  %t/libexec/MacroDefinitionPlugin#MacroDefinition \
 // RUN:   -plugin-path %t/lib/plugins \
```
Comment (bnbarham) [inline test/Macros/macro_plugin_searchorder.swift:66]: Add `-load-plugin-library` here too?

**Review by bnbarham — APPROVED** [review]

**Pull request merged**